linker/resolve.go

```diff
@@ -37,6 +37,15 @@ func (r *result) ResolveMessageType(name protoreflect.FullName) protoreflect.Mes
 	return nil
 }
 
+func (r *result) ResolveOptionsType(name protoreflect.FullName) protoreflect.MessageDescriptor {
+	d, _ := ResolverFromFile(r).FindDescriptorByName(name)
+	md, _ := d.(protoreflect.MessageDescriptor)
+	if md != nil {
+		r.markUsed(md.ParentFile().Path())
```

Comment:
Do we need to check for nil here for `md.ParentFile()`: https://pkg.go.dev/google.golang.org/protobuf/reflect/protoreflect#Descriptor

> Support for this functionality is optional and may return nil.

Reply:
Interesting. That is super strange that they mark parts of the API as optional. The implementation in the Go runtime supports everything (there may actually be more than one, but they all support this method). And I think other implementations are likely going to be rare thanks to [this](https://github.com/protocolbuffers/protobuf-go/blob/master/reflect/protoreflect/type.go#L111).

There _is_ another implementation floating around... [in this repo](https://github.com/bufbuild/protocompile/blob/main/linker/descriptors.go#L530). But it fully implements the interface.

So this will be fine inside of Buf and the BSR. But if someone using this library supplies such a "partial implementation", such as returning one from a `protocompile.Resolver` or by passing one to a direct invocation of `linker.Link`, then it would cause problems for them. I guess I could add a note to both of those methods mentioning that descriptors should implement all methods of the interface, even ones marked optional.

Reply:
I stand corrected. It looks like it's possible for the implementation in the Go runtime to use "placeholders" -- like when initialization of the descriptor fails. Failure can happen due to a variety of issues, but I suspect the most common is a mismatch between a file's "canonical" name (how it is compiled and thus how it is registered in the global registry) and how it is imported. In those cases, the importer will not be able to resolve the dependency descriptor or relevant types and instead use placeholder descriptors, which return nil for some of the methods.

Interesting catch. I will update the guard condition on line 43.

Reply:
Updated this check. Also updated other calls to `ParentFile()` in this repo. Another optional method is `Parent()`, but that one isn't used the same way (only one call that assumes non-nil return, and it's using a concrete impl in this package, so we know it won't return nil).